### PR TITLE
[1.13] Final touches for recipes

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
This PR is focused on final changes to the recipe system to make it more friendly for use outside of crafting.  The main addition is `RecipeWrapper`, which is a reverse version of `InvWrapper` that can be used for recipe matching with item handlers.

There are also a few documentation updates in `ForgeRecipeManager` and `IForgeRecipe`.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java [hunk 1]
```diff
@@ -72,6 +100,7 @@ public ItemStack getResult(IInventory input, World world, RecipeType<?> type)
     /**
      * Modders should not use this if possible.  In the context that this is used, you should already have the recipe.  Do not run extra lookups using this method.
      */
+    @Deprecated
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java:104]: Why?

Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java:104]: Same reason the comment is there.  Method callers should try to reduce recipe lookups as much as possible, this method does not support that.  Nobody should be using it.

File: src/main/java/net/minecraftforge/items/IRecipeInventory.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    default void closeInventory(EntityPlayer player) {}
+    default boolean isItemValidForSlot(int index, ItemStack stack) { return false; }
+    default int getField(int id) { return 0; }
+    default void setField(int id, int value) {}
+    default int getFieldCount() { return 0; }
+    default void clear() {}
+    default ITextComponent getName() { return null; }
+    default boolean hasCustomName() { return false; }
+    default ITextComponent getCustomName() { return null; }
+
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/items/IRecipeInventory.java]: does putting @Override on all of these work? So we know if vanilla IInventory ever shifts.

File: src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java (hunk 1 above)
Comment (liach) [reply src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java:104]: So call the `getRemaingItems` for the recipe object instead. Good suggestion; I hope you can document it.

Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java:104]: Noted to call IRecipe#getRemainingItems in javadoc of that method.

File: src/main/java/net/minecraftforge/items/IRecipeInventory.java (hunk 2 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/items/IRecipeInventory.java]: Overrides have been added.

**Review by liach — APPROVED** [review]

File: src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java [hunk 3]
```diff
@@ -48,20 +54,42 @@ public void onResourceManagerReload(IResourceManager resourceManager) {
         CraftingHelper.reloadConstants(resourceManager);
     }
 
+    /**
+     * Returns all the recipes that match this type.
+     * @param type A recipe type.
+     * @return A mutable view of all the recipes that match this type.  Callers should not modify the returned list.
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java]: `mutable` and `should not modify` are contradictory.
Either just remove mutable or say `immutable but live view`

**Comment by Shadows-of-Fire** [comment]
Switched to a wrapper-based approach instead of another interface that stood alongside IIH.  Hopefully this resolves the concerns which were holding this back.

**Review by tterrag1098 — APPROVED** [review]

**Pull request merged**